dffml/distribuited/orchestrator.py

```diff
@@ -0,0 +1,427 @@
+    CONFIG = NatsWorkerNodeContextConfig
+
+    async def __aenter__(self,):
+        self.uid = secrets.token_hex(nbytes=NBYTES_UID)
+        self.nc = self.parent.nc
+        self.context_done = asyncio.Future()
+        self._stack = AsyncExitStack()
+        await self.init_context()
+        return self
+
+    async def set_instance_details(self, msg):
+        subject = msg.subject
+        # Operation name might contain '.'
+        *subject, onid = subject.split(".")
+        subject = ".".join(subject)
+        operation_name = subject.replace("InstanceDetails", "")
+
+        # data is expected to contain 'instance_name':('operation_name','token')
+        data = msg.data.decode()
+
+        data = json.loads(data)
+        self.logger.debug(f"Received instance details: {data}")
+        token = data["token"]
+        instance_name = data["instance_name"]
+        config = data["config"]
+
+        self.operation_instances[instance_name] = self.operation_names[
+            operation_name
+        ]._replace(instance_name=instance_name)
+        if config:
+            self.operation_configs[instance_name] = config
+
+    async def init_context(self):
+        self.onid = self.config.orchestrator_node_id
+        self.logger.debug(
+            f"New worker node context {self.uid} connected to orchestrator node {self.onid}"
+        )
+
+        self.operation_implemtaions = {}
+        self.operation_configs = {}
+        self.operation_instances: Dict[str, Operation] = {}
+        self.operation_token = {}  # op_name -> token
+
+        self.operation_names = {
+            operation.name: operation
+            for operation in self.parent.config.operations
+        }
+
+        # Subscribe to instance details
+        # We cannot request for instance details because,
+        # all worker nodes need to complete registration before
+        # orchestrator node can allocate instances uniformly
+        for operation in self.parent.config.operations:
+            await self.nc.subscribe(
+                f"InstanceDetails{operation.name}.{self.onid}",
+                queue="SetInstanceDetails",
+                cb=self.set_instance_details,
+            )
+
+        # Send list of operation supported by the node
+        msg_data = json.dumps(list(self.operation_names.keys())).encode()
+
+        # Request node registration
+        self.logger.debug(
+            "Requesting context registration and token allocation"
+        )
+        response = await self.nc.request(
+            f"{RegisterWorkerNode}.{self.onid}", msg_data
+        )
+        # Response contains operation names mapped to token number
+        response = json.loads(response.data.decode())
+        self.operation_token.update(response)
+        self.logger.debug(f"Tokens received: {response}")
+
+        # Look for implementations provided in sub node:
+        for operation_name in response:
+            if operation_name in self.parent.config.implementations:
```

Comment:
The else case here is not handled, what is the else case?

Reply:
If no implementations are provided, the operation is assumed to be entrypoint loadable. opimpNetworkctx raises an exception if an implementation is not found in both ways.